rust/src/core/common/converter.rs

```diff
@@ -0,0 +1,1009 @@
+//! Converter trait and ConverterMap collection.
+//!
+//! Defines the [`Converter`] trait and provides a [`ConverterMap`] collection
+//! to store Converters and provide caching for these Converters.
+//!
+//! [`Converter`]: ./trait.Converter.html
+//! [`ConverterMap`]: ./struct.ConverterMap.html
+
+use crate::common::{
+    error::{inv_arg, oe_err, oe_inv_arg, Result},
+    gates::UnboundGate,
+    types::{ArbData, Gate, Matrix, QubitRef},
+};
+use integer_sqrt::IntegerSquareRoot;
+use num_complex::Complex64;
+use std::{cell::RefCell, collections::HashMap, convert::TryInto, f64::consts::PI, hash::Hash};
+
+/// A type that can be constructed from (part of) an ArbData object.
+///
+/// Any data not consumed by the construction stays in the ArbData object for
+/// further interpretation.
+pub trait FromArb
+where
+    Self: std::marker::Sized,
+{
+    /// Construct from the given ArbData, taking the parameters used for
+    /// construction out of the data object.
+    fn from_arb(src: &mut ArbData) -> Result<Self>;
+}
+
+impl FromArb for () {
+    fn from_arb(_: &mut ArbData) -> Result<Self> {
+        Ok(())
+    }
+}
+
+impl FromArb for u64 {
+    fn from_arb(src: &mut ArbData) -> Result<Self> {
+        let mut args = src.get_args_mut().drain(..1);
```

Comment:
Needs a len check or fails for `ArbData::default()`.

Reply:
Fixed in 45b9a6ac